src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -25,16 +24,21 @@ private sealed class StateSet
             public readonly string Language;
             public readonly DiagnosticAnalyzer Analyzer;
 
-            private readonly ConcurrentDictionary<DocumentId, ActiveFileState> _activeFileStates;
-            private readonly ConcurrentDictionary<ProjectId, ProjectState> _projectStates;
+            private ImmutableDictionary<DocumentId, ActiveFileState> _activeFileStates;
+            private ImmutableDictionary<ProjectId, ProjectState> _projectStates;
+
+            // Used to protect writes of _activeFileStates or _projectStates
+            private readonly object _lock;
 
             public StateSet(string language, DiagnosticAnalyzer analyzer)
             {
                 Language = language;
                 Analyzer = analyzer;
 
-                _activeFileStates = new ConcurrentDictionary<DocumentId, ActiveFileState>(concurrencyLevel: 2, capacity: 10);
-                _projectStates = new ConcurrentDictionary<ProjectId, ProjectState>(concurrencyLevel: 2, capacity: 1);
+                _activeFileStates = ImmutableDictionary<DocumentId, ActiveFileState>.Empty;
+                _projectStates = ImmutableDictionary<ProjectId, ProjectState>.Empty;
+
+                _lock = new object();
```

Comment:
nit: my preference would be field initializers.

Reply:
Created one for the lock, left the others the way the are/were.